// @elastic/react-search-ui/lib/containers/Sorting.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SortingContainer = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = require("react");

var _ = require("..");

var _reactSearchUiViews = require("@elastic/react-search-ui-views");

var _SortList = _interopRequireDefault(require("../types/SortList"));

var _types = require("../types");

var _excluded = ["className", "label", "setSort", "sortDirection", "sortField", "sortList", "sortOptions", "view"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function findSortOption(sortOptions, sortData) {
  if (sortData.indexOf("|||") === -1) {
    return sortOptions.find(function (option) {
      return JSON.stringify(option.value) === sortData;
    });
  }

  var _sortData$split = sortData.split("|||"),
      _sortData$split2 = (0, _slicedToArray2.default)(_sortData$split, 2),
      value = _sortData$split2[0],
      direction = _sortData$split2[1];

  return sortOptions.find(function (option) {
    return option.value === value && option.direction === direction;
  });
}

function formatValue(sortField, sortDirection, sortList) {
  if (sortList && sortList.length > 0) {
    return JSON.stringify(sortList);
  }

  return "".concat(sortField, "|||").concat(sortDirection);
}

function formatSelectValues(sortOption) {
  if (Array.isArray(sortOption.value)) {
    // save value as string for comparison
    return JSON.stringify(sortOption.value);
  } else {
    return formatValue(sortOption.value, sortOption.direction);
  }
}

function formatSelectOption(sortOption) {
  return {
    label: sortOption.name,
    value: formatSelectValues(sortOption)
  };
}

var SortingContainer = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(SortingContainer, _Component);

  var _super = _createSuper(SortingContainer);

  function SortingContainer() {
    (0, _classCallCheck2.default)(this, SortingContainer);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(SortingContainer, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          label = _this$props.label,
          setSort = _this$props.setSort,
          sortDirection = _this$props.sortDirection,
          sortField = _this$props.sortField,
          sortList = _this$props.sortList,
          sortOptions = _this$props.sortOptions,
          view = _this$props.view,
          rest = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
      var View = view || _reactSearchUiViews.Sorting;
      return View(_objectSpread({
        className: className,
        label: label,
        onChange: function onChange(o) {
          var sortOption = findSortOption(sortOptions, o);
          setSort(sortOption.value, sortOption.direction);
        },
        options: sortOptions.map(formatSelectOption),
        value: formatValue(sortField, sortDirection, sortList)
      }, rest));
    }
  }]);
  return SortingContainer;
}(_react.Component);

exports.SortingContainer = SortingContainer;
(0, _defineProperty2.default)(SortingContainer, "propTypes", {
  // Props
  className: _propTypes.default.string,
  label: _propTypes.default.string,
  sortOptions: _propTypes.default.arrayOf(_types.SortOption).isRequired,
  view: _propTypes.default.func,
  // State
  sortDirection: _propTypes.default.oneOf(["asc", "desc", ""]),
  sortField: _propTypes.default.string,
  sortList: _SortList.default,
  // Actions
  setSort: _propTypes.default.func.isRequired
});

var _default = (0, _.withSearch)(function (_ref) {
  var sortDirection = _ref.sortDirection,
      sortField = _ref.sortField,
      sortList = _ref.sortList,
      setSort = _ref.setSort;
  return {
    sortDirection: sortDirection,
    sortField: sortField,
    sortList: sortList,
    setSort: setSort
  };
})(SortingContainer);

exports.default = _default;
