// @elastic/react-search-ui-views/lib/PagingInfo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _viewHelpers = require("./view-helpers");

var _excluded = ["className", "end", "searchTerm", "start", "totalResults"];

function PagingInfo(_ref) {
  var className = _ref.className,
      end = _ref.end,
      searchTerm = _ref.searchTerm,
      start = _ref.start,
      totalResults = _ref.totalResults,
      rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    className: (0, _viewHelpers.appendClassName)("sui-paging-info", className)
  }, rest), "Showing", " ", /*#__PURE__*/_react.default.createElement("strong", null, start, " - ", end), " ", "out of ", /*#__PURE__*/_react.default.createElement("strong", null, totalResults), searchTerm && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, " ", "for: ", /*#__PURE__*/_react.default.createElement("em", null, searchTerm)));
}

PagingInfo.propTypes = {
  end: _propTypes.default.number.isRequired,
  searchTerm: _propTypes.default.string.isRequired,
  start: _propTypes.default.number.isRequired,
  totalResults: _propTypes.default.number.isRequired,
  className: _propTypes.default.string
};
var _default = PagingInfo;
exports.default = _default;
