// @elastic/react-search-ui-views/lib/Paging.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _rcPagination = _interopRequireDefault(require("rc-pagination"));

var _en_US = _interopRequireDefault(require("rc-pagination/lib/locale/en_US"));

var _viewHelpers = require("./view-helpers");

var _excluded = ["className", "current", "resultsPerPage", "onChange", "totalPages"];

function Paging(_ref) {
  var className = _ref.className,
      current = _ref.current,
      resultsPerPage = _ref.resultsPerPage,
      onChange = _ref.onChange,
      totalPages = _ref.totalPages,
      rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement(_rcPagination.default, (0, _extends2.default)({
    current: current,
    onChange: onChange,
    pageSize: resultsPerPage,
    total: totalPages * resultsPerPage,
    className: (0, _viewHelpers.appendClassName)("sui-paging", className),
    locale: _en_US.default
  }, rest));
}

Paging.propTypes = {
  current: _propTypes.default.number.isRequired,
  onChange: _propTypes.default.func.isRequired,
  resultsPerPage: _propTypes.default.number.isRequired,
  totalPages: _propTypes.default.number.isRequired,
  className: _propTypes.default.string
};
var _default = Paging;
exports.default = _default;
